ros/content.md

```diff
@@ -1,28 +1,99 @@
 
 > [wikipedia.org/wiki/Robot_Operating_System](https://en.wikipedia.org/wiki/Robot_Operating_System)
 
-[%%LOGO%%](http://www.ros.org/)
+[%%LOGO%%](https://index.ros.org/doc/ros2)
 
 # How to use this image
 
-## Create a `Dockerfile` in your ROS app project
+## Creating a `Dockerfile` to install ROS packages
+
+To create your own ROS docker images and install custom packages, here's a simple example of installing the C++, Python client library demos and security CLI using the official released Debian packages via apt-get.
 
 ```dockerfile
-FROM %%IMAGE%%:indigo
-# place here your application's setup specifics
-CMD [ "roslaunch", "my-ros-app my-ros-app.launch" ]
+FROM %%IMAGE%%:dashing
+
+# install ros package
+RUN apt-get update && apt-get install -y \
+      ros-${ROS_DISTRO}-demo-nodes-cpp \
+      ros-${ROS_DISTRO}-demo-nodes-py && \
+    rm -rf /var/lib/apt/lists/*
+
+# launch ros package
+CMD ["ros2", "launch", "demo_nodes_cpp", "talker_listener.launch.py"]
 ```
 
-You can then build and run the Docker image:
+Note: all ROS images include a default entrypoint that sources the ROS environment setup before executing the configured command, in this case the demo packages launch file. You can then build and run the Docker image like so:
 
 ```console
-$ docker build -t my-ros-app .
-$ docker run -it --rm --name my-running-app my-ros-app
+$ docker build -t my/ros:app .
+$ docker run -it --rm my/ros:app
+[INFO] [launch]: process[talker-1]: started with pid [813]
+[INFO] [launch]: process[listener-2]: started with pid [814]
+[INFO] [talker]: Publishing: 'Hello World: 1'
+[INFO] [listener]: I heard: [Hello World: 1]
+[INFO] [talker]: Publishing: 'Hello World: 2'
+[INFO] [listener]: I heard: [Hello World: 2]
+...
+```
+
+## Creating a `Dockerfile` to build ROS packages
+
+To create your own ROS docker images and build custom packages, here's a simple example of installing a package's build dependencies, compiling it from source, and installing the resulting build artifacts into a final multi-stage image layer.
+
+```dockerfile
+FROM %%IMAGE%%:dashing-ros-base AS builder
+
+# clone package source
+ENV ROS_WS /opt/ros_ws
+RUN mkdir -p $ROS_WS/src
+WORKDIR $ROS_WS
+RUN git -C src clone \
+      -b $ROS_DISTRO \
+      https://github.com/ros2/demos.git
+
+# install package dependencies
+RUN . /opt/ros/$ROS_DISTRO/setup.sh && \
+    apt-get update && rosdep update && \
+    rosdep install -y \
+      --from-paths \
+        src/demos/demo_nodes_cpp \
+      --ignore-src && \
+    rm -rf /var/lib/apt/lists/*
+
+# build package workspace
+RUN . /opt/ros/$ROS_DISTRO/setup.sh && \
+    colcon build \
+      --packages-select \
+        demo_nodes_cpp \
+      --mixin release
+
+# create new multi-stage
+FROM %%IMAGE%%:dashing-ros-core AS runner
```

Comment:
Could we end up with missing runtme dependencies by having the runner being based on a different image than the builder?
Maybe the runner should use rosdep to install runtime dependencies?

Reply:
@mikaelarguedas' question is actually partly why I posted [Generating ‘dev’ and runtime artefacts from ROS packages](https://discourse.ros.org/t/generating-dev-and-runtime-artefacts-from-ros-packages/12448).

Reply:
Looks like using rosdep to install runtime dependencies has been ticketed here:
https://github.com/ros-infrastructure/rosdep/pull/727

Once rosdep gets that feature, we could update example to use it. In the meantime, we'll just make a note to make sure the user is aware of need for runtime depencicies.

Reply:
Doesn't this basically mean that these `Dockerfile`s are currently broken? Without the runtime dependencies, what's in `install` most likely doesn't run in the `runner` container.

Reply:
> Once rosdep gets that feature, we could update example to use it. In the meantime, we'll just make a note to make sure the user is aware of need for runtime depencicies.

IMO this doesnt justify not installing runtime dependencies in the runner image. So I would argue that rosdep install should be used in the runner image and once rosdep allows to restrict the type of dependency to install, we can update the example to install "only runtime" dependencies.

These example dockerfiles work but only out of luck because the talker listener happen to have all their dependencies satisfied by `ros_core`. This will most likely not be the case for most use cases making this example confusing for users.

---

This also brings another point that is not explained here that is "why use ros-base a the builder base when ros-core is enough and consistent with the runner?"
This needs some comment explaining that the ros-core image doesnt come with colcon and other build tools anymore so ros_base needs to be used instead.

Reply:
> This needs some comment explaining that the ros-core image doesnt come with colcon and other build tools anymore so ros_base needs to be used instead.

Yep, this'll need to be updated.

> These example dockerfiles work but only out of luck because the talker listener happen to have all their dependencies satisfied by ros_core. This will most likely not be the case for most use cases making this example confusing for users.

What you alternative example package would you recommend? I only chose the talker and listener example as it's the classic demo, and its simple/quick to build, launch, and verify.

Reply:
I think it's good to keep the `demo_nodes_cpp` package as the example :+1:

A good way to illustrate the need for installing runtime dependencies in the runner would be to launch the `add_two_ints.launch.py` example instead. As that uses insterfaces from `example_interfaces` that are not in ros_core, the container will fail.

Reply:
In other words here is the runner I'm suggesting:

```
# create new multi-stage
FROM %%IMAGE%%:dashing-ros-core-bionic AS runner

# copy package install
ENV ROS_WS /opt/ros_ws
COPY --from=builder  $ROS_WS/install $ROS_WS/install

# install package dependencies
RUN apt-get update && apt-get install -y python3-rosdep && \
    apt-mark manual sudo && \
    rosdep init && \
    rosdep update --rosdistro $ROS_DISTRO && \
    . $ROS_WS/install/setup.sh && \
    rosdep install -y \
      --from-paths \
        $ROS_WS/install/share \
      --ignore-src && \
    apt-get remove -y python3-rosdep && apt-get autoremove -y && \
    rm -rf /var/lib/apt/lists/*

# source entrypoint setup
RUN sed --in-place --expression \
      '$isource "$ROS_WS/install/setup.bash"' \
      /ros_entrypoint.sh
# run launch file
CMD ["ros2", "launch", "demo_nodes_cpp", "add_two_ints.launch.py"]
```

Reply:
I think having to install rosdep back into ros-core is a little counter to its purpose. If rosdep could pipe a list of package names to a files that satisfy runtime requirements, then that file could be copied and piped back into apt-get in the runner image.

Although, this is still blocked by rosdep's feature set, so until then I've updated the docs to make careful note of this in 66047eb

Reply:
> this is still blocked by rosdep's feature set

the rosdep feature set allows to do it. It's just not as optimized as it could be and we can easily add a flag once there is one to restrict the dependency set while preserving the logic of "automatically installing the dependencies you need" rather than having to figure it out and install them by hand.

I still think that this example is incorrect without it, either you're lucky and your dependencies are already in the base image, at which point rosdep is not necessary in the builder image either. Or (what I would guess is the vast majority of the cases) you need to install extra dependencies and ROS provides a tool for that called rosdep that we should use.

So I'm still in favor of using rosdep in the runner image as of now, uninstalling rosdep after was just an extra cleanup that may or may not be justified as rosdep has a pretty small footprint.

> If rosdep could pipe a list of package names to a files that satisfy runtime requirements, then that file could be copied and piped back into apt-get in the runner image.

As you use a different base image for building and running, the list of missing dependencies to install will differ between the builder or the runner so it doesnt seem bulletproof.

> helping to reconcile runtime vs development dependency discrepancies you may have otherwise forgotten to declare in your `package.xml` manifest

If the package.xml is not used to install dependencies, this approach cannot help catch missing or mismatching dependency declarations.

Reply:
> As you use a different base image for building and running, the list of missing dependencies to install will differ between the builder or the runner so it doesnt seem bulletproof.

That's where I think there should be a dry run option for rosdep where it just returns all the resolved runtime dependencies, regardless if they are installed or not, which would be complete enough for the runner to apt install. We do this for the CI in the navigation stack, but with a few more POSIX tools.

> If the package.xml is not used to install dependencies, this approach cannot help catch missing or mismatching dependency declarations.

That out of band dependency management then just falls onto the developer doesn't it?
rosdep can't help the developer if the developer doesn't also help rosdep.

Reply:
> That out of band dependency management then just falls onto the developer doesn't it?
rosdep can't help the developer if the developer doesn't also help rosdep.

That's why I'm arguing for using rosdep in the example :wink: 
Otherwise this statement is misleading because the example does *not* allow catching mismatching dependency declarations because it ignores the declared dependencies (by not using rosdep)

> That's where I think there should be a dry run option for rosdep where it just returns all the resolved runtime dependencies

There is currently a dry run option, that lists all the installation commands it would run, (but it only lists the missing packages).
Is there an issue on rosdep to do the same for all (even already installed) packages ? it doesnt look like it would be a difficult feature to add.

Reply:
I don't know of such a ticket, but yeah, it should only need a small change to support. Then we can pipe those apt package names to a placeholder file to share across multistage.

Reply:
> Then we can pipe those apt package names

I think piping the list of commands is better than the apt package names. Otherwise how would you handle other installers (e.g. pip packages)? trying to apt install those will likely not work.

> I don't know of such a ticket

I encourage you to open one if that's a feature you need and end up working around in other places.

In the meantime, can we agree on updating the example to install dependencies in the runner to move this forward? once extra rosdep features are available we can see how to update/optimize this example

Reply:
The `--simulate` option of `rosdep` could *almost* do what you are asking for I believe.

See also [How can generate a list of dependencies from the src folder of a workspace?](https://answers.ros.org/question/251732) on ROS Answers.

Reply:
:+1: yes that's the one I had in mind in:

> There is currently a dry run option, that lists all the installation commands it would run, (but it only lists the missing packages).

Reply:
> I encourage you to open one if that's a feature you need

On a closer search, looks like it's been open for while:

https://github.com/ros-infrastructure/rosdep/issues/587
https://github.com/ros-infrastructure/rosdep/pull/727

I've changed the multistage example to demo build and dependency caching with 2316f47 . We can always append a runtime stage later when rosdep is better up to the task. I think it best to keep the advance example clean and development oriented, so we can leave it at the builder stage to install exec dependencies for now.